qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,849 @@
+
+        if clsname not in ConstraintsDescriptor.global_constraints:
+            gendef = getattr(owner, "_define_constraints")
+            source = gendef()
+            self.add_new_constraints(clsname, owner.PARAM_DEF, *source)
+
+        return ConstraintsDescriptor.global_constraints[clsname]
+
+    @classmethod
+    def add_new_constraints(cls, pulse_name: str, params: List[str], *exprs: Union["Expr", List]):
+        """Add new pulse parameter constraints to descriptor.
+
+        Args:
+            pulse_name: Name of pulse subclass.
+            params: List of parameter names.
+            exprs: Symbolic expressions to validate assigned parameters.
+        """
+        import sympy as sym
+
+        params = [sym.Symbol(p) for p in ["limit"] + params]
+        constraints = []
+        for expr in exprs:
+            if isinstance(expr, list):
+                # If clause
+                constraints.append(_lambdify_ite(expr, params))
+            else:
+                constraints.append(sym.lambdify(params, expr))
+        cls.global_constraints[pulse_name] = constraints
+
+
+class SymbolicPulse(Pulse):
+    """The abstract superclass for parametric pulses with symbolic expression.
+
+    A symbolic pulse subclass can be defined with an envelope and parameter constraints.
+    The parameters constitute the waveform should be defined in the class attribute
+    :attr:`SymbolicPulse.PARAM_DEF`. This must include ``duration`` as a first element
+    to define the length of pulse instruction for scheduling.
+    For now, envelope and constraints must be defined with SymPy expressions.
+
+    .. rubric:: Envelope function
+
+    This is defined with a class method :meth:`_define_envelope` that returns
+    a single symbolic expression which is a function of ``t`` and ``duration``.
+    The time ``t`` and ``duration`` are in units of dt, i.e. system cycle time,
+    and this function is sampled with a discrete time vector in [0, ``duration``]
+    sampling the pulse envelope at every 0.5 dt (middle sampling strategy) when
+    :meth:`get_waveform` method is called.
+    The function can generate complex-valued samples representing a phase in Gaussian plane,
+    and real and imaginary part may be supplied to two quadratures of
+    the IQ mixer in the control electronics.
+    The sample data is not generated until the get method is called thus
+    a symbolic pulse instance only stores parameter values and waveform shape,
+    which greatly reduces memory footprint during the program generation.
+
+
+    .. rubric:: Constraint functions
+
+    This is defined with a class method :meth:`_define_constraints` that returns
+    a list of symbolic expressions. These constraint functions are called
+    when a symbolic pulse instance is created with assigned parameter values with ``limit``
+    representing the policy of allowing the maximum amplitude exceeding 1.0
+    (defaults to ``True``, i.e. it doesn't allow amplitude > 1.0).
```

Comment:
```suggestion
    a list of symbolic expressions. These constraint functions are called
    when a symbolic pulse instance is created with assigned parameter values. The ``limit`` variable
    represents the policy of allowing the maximum amplitude to exceed 1.0
    (its default value is ``True``, i.e. it doesn't allow amplitude > 1.0).
```
is `variable` the correct word? Should it be `argument` instead?

Reply:
variable should be okey from sympy perspective